Issue 63: Implementing SBC

SamuelBrand1 opened the pull request:
~This is an ongoing PR for implementing SBC.~

- [x] plots.
- [x] prior predictive sampling over a range.
- [x] Posterior sample for each prior predictive.
- [x] Unit tests.
- [x] Example usage.

This PR is ready for review now. It implements SBC methods very closely modelled on https://github.com/arviz-devs/simulation_based_calibration but for `numpyro` models.

For more extensive explanation see the committed notebook which covers the theoretical basis for SBC as well as an example implementation.

~NB: the problem with pre-commit which #67 was aimed at is obviously not resolved.~

This PR closes #63

AFg6K7h4fhy2 commented:
Putting these images for others to see, if interested:

<details markdown=1>

<summary>
Histogram plot of SBC results
</summary>

<img width="515" alt="Screenshot 2025-02-11 at 14 20 18" src="https://github.com/user-attachments/assets/1713031d-f23c-4954-a504-a73d16a291d6" />

</details>



<details markdown=1>

<summary>
ECDF/PIT plot of SBC results
</summary>

<img width="513" alt="Screenshot 2025-02-11 at 14 21 34" src="https://github.com/user-attachments/assets/07fb2dd5-b1f6-4f0e-af66-f7b986715acf" />

</details>

AFg6K7h4fhy2 commented on forecasttools/sbc.py:
```diff
@@ -0,0 +1,206 @@
+from jax import random
+from numpyro.infer import MCMC
+from numpyro.infer.mcmc import MCMCKernel
+from tqdm import tqdm
```
AI summary from DuckDuckGo: _"Tqdm is a Python library that provides a fast and easy way to create progress bars for loops and iterables, allowing users to visualize the progress of their code execution"_.

Neat addition, might use myself. 

Not sure if the best choice but worth remarking on.

AFg6K7h4fhy2 commented on forecasttools/sbc.py:
```diff
@@ -0,0 +1,206 @@
+            initial=self._simulations_complete,
+            total=self.num_simulations,
+        )
+        try:
```
Were there any errors that you got (or were looking out for) when initially composing this, possibly before the `try` block usage? (asking out of curiosity).

SamuelBrand1 replied:
No.

SamuelBrand1 replied:
Getting the shape of the rank statistics was the main problem.

AFg6K7h4fhy2 commented on forecasttools/sbc.py:
```diff
@@ -0,0 +1,206 @@
+        idata = az.from_numpyro(mcmc)
+        return idata, num_samples
+
+    def run_simulations(self):
```
```suggestion
    def run_simulations(self) -> None:
```

(if choosing to follow theme of hinting)

AFg6K7h4fhy2 commented on tests/test_sbc.py:
Maybe something for 

* rank uniformity
* empty observed variables

would be good additions here. 

I will address these again more on 2025-02-12.

SamuelBrand1 replied:
Could you be more specific?

AFg6K7h4fhy2 commented on notebooks/sbc_model_checking.qmd:
* Verifying (2025-02-11) on this that I was able to run everything successfully, without error or warning, and that I added the plots as a comment on this PR.
* My preference (free to disregard, here) would be for the section-titles to have the first letter of each word capitalized.
* Wikipedia links added to certain terms here might be useful but also might not be worth the time.

AFg6K7h4fhy2 commented on notebooks/sbc_model_checking.qmd:
```diff
@@ -0,0 +1,184 @@
+-   Tuning and running SBC for a `numpyro` model using the `SBC` from `forecasttools`.
+-   Plotting results and interpreting them.
+
+The example we will use to illustrate our ideas is the classic "eight schools" inference problem from section 5.5 of Gelman *et al* [@gelman2013] with posterior sampling done using the NUTS algorithm implemented by `numpyro` PPL.
+
```
```suggestion
The example we will use to illustrate our ideas is the classic "eight schools" inference problem from section 5.5 of Gelman *et al* [@gelman2013] with posterior sampling done using the NUTS algorithm implemented by `numpyro` PPL[^eight_schools_note].

[^eight_schools_note]: The interested reader can find a worked example of the eight schools problem on NumPyro's homepage (see [here](https://num.pyro.ai/en/stable/getting_started.html)), along with other examples, such as _Using NUTS for a simple hierarchical model_ (see [here](https://github.com/pyro-ppl/numpyro/blob/master/examples/baseball.py)). 
```

AFg6K7h4fhy2 commented on notebooks/sbc_model_checking.qmd:
```diff
@@ -0,0 +1,184 @@
+F^{(-1)}(\theta[k]) \sim \mathcal{U}(0,1).
+$$
+
+Where $F^{(-1)}$ is the (pseudo) inverse distribution function of the known distribution $p(\theta[k])$. This is often call ed the probability integral transform (PIT). Significant (in a classical statistics sense) deviations in the PIT of parameters from the uniform distribution form evidence that either the model is incorrectly coded, its priors are cover hard to sample regimes, and/or the Baye ...
```
```suggestion
Where $F^{(-1)}$ is the (pseudo) inverse distribution function of the known distribution $p(\theta[k])$. This is often called the probability integral transform (PIT). Significant (in a classical statistics sense) deviations in the PIT of parameters from the uniform distribution form evidence that either the model is incorrectly coded, its priors are cover hard to sample regimes, and/or the Bayesian inference method is approximate/invalid. See Talts *et al* [@talts] for more details on understanding the returned tests.
```

Grammar edit.

AFg6K7h4fhy2 commented on notebooks/sbc_model_checking.qmd:
```diff
@@ -0,0 +1,184 @@
+
+## Simulation-based calibration
+
+"Unit" testing the correctness of Bayesian probabilistic programs is challenging because the target quantity is a distribution (the posterior distribution) typically sampled from via a random process, e.g. Markov chain Monte Carlo sampling. The range of models where the correct posterior distribution is known analytically, and therefore, we can test sampling against distributional checks (e.g. Ko ...
```
```suggestion
Unit testing the correctness of Bayesian probabilistic programs is challenging because the target quantity is a distribution (the posterior distribution), which is typically sampled via a random process, such as Markov chain Monte Carlo sampling. The range of models for which the correct posterior distribution is known analytically is limited, making it difficult to test sampling methods against distributional checks (e.g., Kolmogorov-Smirnov tests, chi-squared tests, etc.).
```

Had trouble reading this first time over. 

Perhaps the change (of which some of it is stylistic in nature) I've suggested is an improvement.

AFg6K7h4fhy2 reviewed (changes requested):
* Thank you very much for this PR @SamuelBrand1 !
* I have left some initial comments.
* I would like to critically go through some more of the code in `sbc_model_checking.qmd` and the tests.
* Most of the comments are lower priority but nice to have.
* Rank-uniformity checks + variables constitute the only moderately impactful suggestion, in my mind.
* The notebook is quite nice (as mentioned in private communication).
* I will try to respond promptly to an return remarks you have to mine.
* Some of my comments can be resolved after reading, others can be resolved once you've stated your preference and you believe I've read the comment.

SamuelBrand1 commented on forecasttools/sbc_plots.py:
```diff
@@ -0,0 +1,144 @@
+
+    if n_plots > 1:
+        if figsize is None:
+            figsize = (8, n_plots * 1.0)
```
Actually I think `float` works better here. The original was from https://github.com/arviz-devs/simuk/blob/5f8454c37348a7c301f36a4bb8fec2ce81e117ef/simulation_based_calibration/plots.py#L54 which uses `0.75`. That rendered as too squashed

AFg6K7h4fhy2 replied:
sounds good!

SamuelBrand1 commented on forecasttools/sbc_plots.py:
```diff
@@ -0,0 +1,144 @@
+
+    Parameters
+    ----------
+    simulations : dict[str] -> listlike
```
Yes, this is not really good type hinting here. I'll change this

SamuelBrand1 replied:
Changed

SamuelBrand1 commented on forecasttools/sbc.py:
```diff
@@ -0,0 +1,206 @@
+        self.kwargs = kwargs
+        self.observed_vars = observed_vars
+
+        for key in self.observed_vars:
+            if key in self.kwargs and self.kwargs[key] is not None:
+                raise ValueError(
+                    f"The value for '{key}' in kwargs must be None for this to"
+                    " be a prior predictive check."
+                )
```
Automation is better unless you've got a counter-example?

SamuelBrand1 commented on forecasttools/sbc.py:
```diff
@@ -0,0 +1,206 @@
+                        " consistent."
+                    )
+                posterior = idata["posterior"]
+                for name in prior:
+                    num_dims = jnp.ndim(prior_draw[name])
+                    if num_dims == 0:
+                        self.simulations[name].append(
+                            (posterior[name].sel(chain=0) < prior_draw[name])
+                            .sum()
+                            .values
+                        )
+                    else:
+                        self.simulations[name].append(
+                            (posterior[name].sel(chain=0) < prior_draw[name])
+                            .sum(axis=0)
+                            .values
+                        )
```
sure

SamuelBrand1 commented on forecasttools/sbc_plots.py:
```diff
@@ -0,0 +1,144 @@
+    if var_names is None:
+        var_names = list(simulations.keys())
+
+    sims = {}
```
I've stolen the logic here from the pymc/arviz people. It has an expand out process that is neat, and I think works already.

SamuelBrand1 commented on forecasttools/sbc_plots.py:
```diff
@@ -0,0 +1,144 @@
+            ax.set_yticks([])
+            idx += 1
+
+    for j in range(n_plots, len(axes)):
+        fig.delaxes(axes[j])
```
Ok.

SamuelBrand1 commented on forecasttools/sbc_plots.py:
```diff
@@ -0,0 +1,144 @@
+        np.linspace(0, max_rank, len_bins),
+        band[0],
+        band[2],
+        color="0.5",
```
I'd rather keep it simple.

AFg6K7h4fhy2 replied:
Sounds good, not necessary and can always add later.

SamuelBrand1 commented on forecasttools/sbc_plots.py:
```diff
@@ -0,0 +1,144 @@
+    Parameters
+    ----------
+    simulations : dict[str] -> listlike
+        The SBC.simulations dictionary.
```
Changed this!

dylanhmorris commented on forecasttools/sbc.py:
```diff
@@ -0,0 +1,230 @@
+            total=self.num_simulations,
+        )
+        try:
+            while self._simulations_complete < self.num_simulations:
```
Why a `while` rather than a `for` loop here, given that afaict there's no way for an iteration _not_ to increment the counter?

Also, perhaps worth having a function for the loop step?

SamuelBrand1 replied:
Open admission, I started this as a straight grab from https://github.com/arviz-devs/simuk and modified as required. This is a leftover, but happy to keep modifying.

SamuelBrand1 replied:
I've noticed the reason now. When one simulation fails this returns out the gathered SBC trials that did run (see L191). Which is neat. But agree that its cleaner to have an inner loop function

dylanhmorris replied:
Can still wrap a `for`-loop in a try/finally though, no?

SamuelBrand1 replied:
OK, so I assumed while had a nicer cache out behaviour.

SamuelBrand1 replied:
But I've found no evidence of that

SamuelBrand1 replied:
I've refactored the inner loop here with another SBC method that increments the rank statistic for each variable.

dylanhmorris commented on forecasttools/sbc.py:
```diff
@@ -0,0 +1,230 @@
+        idata = az.from_numpyro(mcmc)
+        return idata, num_samples
+
+    def run_simulations(self) -> None:
```
Option to hang on to more verbose (even complete?) values from each run. Probably shouldn't be the default for reasons of memory, but might be useful for inspection, debugging, etc.

SamuelBrand1 replied:
Added an `inspection_mode` (default False) for the SBC instance constructor. When on then the priors, prior preds and the idata objects of each fit are cached into the SBC object.

dylanhmorris commented on forecasttools/sbc_plots.py:
```diff
@@ -0,0 +1,182 @@
+    )
+
+
+class UniformCDF:
```
Why not just use `numpyro.distributions.Uniform` since we depend on numpyro anyway?

SamuelBrand1 replied:
Good shout.

SamuelBrand1 replied:
Done this

dylanhmorris reviewed (approved):
A few minor comments/questions, but thins is looking very good!

SamuelBrand1 commented:
> A few minor comments/questions, but this is looking very good! Thanks, @SamuelBrand1!

Thanks! 

All of your comments are leftovers from me grabbing a pymc/bambi implementation https://github.com/arviz-devs/simuk . I've put an issue up over there https://github.com/arviz-devs/simuk/issues/19 so in the future we might just be importing their package but I hope this implementation works for now.

Note that: a quick implementation in julia I did https://github.com/SamuelBrand1/SBC returns the p-value of the ChiSq uniformity test on the rank statistics. This could be a handy issue/next-step to go along with the viz.

SamuelBrand1 commented:
Following @dylanhmorris 's comments:

- Refactored the inner loop of the SBC. This is nicer, I've left it as a set of method calls because I find that easier to read but YMMV. Having a method to gather rank statistics makes this much cleaner to read.
- Added an `inspection_mode` boolean to the `SBC` class. If `True` then running SBC also saves the prior draws/preds and the `idata` objects of the fit (e.g. MCMC chains). The idea is that you can use this for inspection on a small number of SBC trials to get a sense of whats going on.
- @dylanhmorris spotted a hand rolled discrete uniform was not needed so I've removed and added the `numpyro.distributions` one.

AFg6K7h4fhy2 reviewed (approved):
* Thank you @dylanhmorris for your comments on this PR!
* Thank you again @SamuelBrand1 for this work!
* The tests and notebook both run successfully for me locally.
* There is nothing else within scope that I believe needs to be added for this PR to be merged.
* I approve this PR.
